lib/matplotlib/axis.py

```diff
@@ -1823,9 +1823,12 @@ def set_ticks(self, ticks, labels=None, *, minor=False, **kwargs):
         Parameters
         ----------
         ticks : list of floats
-            List of tick locations.
+            List of tick locations.  The axis `.Locator` is replaced by a
+            `.FixedLocator`.  Some default tick formatters will not label
+            arbitrary ticks, so pass *label* to ensure labels are visible.
```

Comment:
I don't think we should add that second sentence. It's a workaround for another issue. Instead we should

- Document at the respective formatters that they don't generate labels for certain positions.
- In theory that would be enough. If people use a `LogFormatterSciNotation` they should refer to its docs if it does not format as expected. The issue is that `plt.loglog` implicitly switches to this formatter, so people won't know. One should investigate whether empty labels can also occur at positions issued from the corresponding formatter. If not, only special user-defined positions will run into the empty-label issue, which means we can issue a warning on empty labels without annoying regular users.

Reply:
> Document at the respective formatters that they don't generate labels for certain positions.

Implementing your suggestions goes a bit further than the very small PR that I started from - should e.g. adding docs to `Formatter`s be continued here or split off into some other PR?

> In theory that would be enough. If people use a `LogFormatterSciNotation` they should refer to its docs if it does not format as expected. The issue is that `plt.loglog` implicitly switches to this formatter, so people won't know. 

Yes, this implicit `Formatter` switching is the gotcha here, since it breaks the behaviour that the user (ignorant of the internals) would expect - namely *I set the ticks, so I should get labels on them*, which got me here in the first place.

Reply:
> One should investigate whether empty labels can also occur at positions issued from the corresponding formatter.

Regarding the problem of `Locator`-`Formatter` correspondence (I hope I am understanding you correctly): from the issue discussion and my (limited) knowledge of the code, I grasp that is a "gentlemen's agreement" without any explicit mechanism ensuring `Locator` will not generate ticks that `Formatter` will not make labels for.

Digging through the code, I found that the choice of `Locator` and `Formatter` pairs is in the `set_default_locators_and_formatters` override methods of `ScaleBase` subclasses (in scale.py). Do you mean going through all possible combinations and analyzing their interactions...?

> If not, only special user-defined positions will run into the empty-label issue, which means we can issue a warning on empty labels without annoying regular users.

I am not certain what you mean by "we can issue a warning on empty labels without annoying regular users." - in this docstring, that this is only relevant when setting ticks manually? Or somewhere else, or at runtime (?)

Reply:
> I don't think we should add that second sentence. It's a workaround for another issue. Instead we should
> 
> * Document at the respective formatters that they don't generate labels for certain positions.
> * In theory that would be enough. If people use a `LogFormatterSciNotation` they should refer to its docs if it does not format as expected. The issue is that `plt.loglog` implicitly switches to this formatter, so people won't know. One should investigate whether empty labels can also occur at positions issued from the corresponding formatter. If not, only special user-defined positions will run into the empty-label issue, which means we can issue a warning on empty labels without annoying regular users.

You can issue a runtime warning if you want, but I don't see that adding a sentence that says what to do if you don't get the expected results is a bad thing to do.   The heuristics for such a runtime warning seem brittle to me.